src/SDiagonal.jl

```diff
@@ -12,105 +12,59 @@ end
 end
 
-struct SDiagonal{N,T} <: StaticMatrix{N,N,T}
-    diag::SVector{N,T}
-    SDiagonal{N,T}(diag::SVector{N,T}) where {N,T} = new(diag)
-end
-diagtype(::Type{SDiagonal{N,T}}) where {N, T} = SVector{N,T}
-diagtype(::Type{SDiagonal{N}}) where {N} = SVector{N}
-diagtype(::Type{SDiagonal}) = SVector
+const SDiagonal = Diagonal{T,SVector{N,T}} where {N,T}
+SDiagonal(x...) = Diagonal(SVector(x...))
 
 # this is to deal with convert.jl
-@inline (::Type{SD})(a::AbstractVector) where {SD <: SDiagonal} = SDiagonal(convert(diagtype(SD), a))
-@inline (::Type{SD})(a::Tuple) where {SD <: SDiagonal} = SDiagonal(convert(diagtype(SD), a))
-@inline SDiagonal(a::SVector{N,T}) where {N,T} = SDiagonal{N,T}(a)
-
-@generated function SDiagonal(a::StaticMatrix{N,N,T}) where {N,T}
-    expr = [:(a[$i,$i]) for i=1:N]
-    :(SDiagonal{N,T}($(expr...)))
-end
-
-convert(::Type{SDiagonal{N,T}}, D::SDiagonal{N,T}) where {N,T} = D
-convert(::Type{SDiagonal{N,T}}, D::SDiagonal{N}) where {N,T} = SDiagonal{N,T}(convert(SVector{N,T}, D.diag))
-
-function getindex(D::SDiagonal{N,T}, i::Int, j::Int) where {N,T}
-    @boundscheck checkbounds(D, i, j)
-    @inbounds return ifelse(i == j, D.diag[i], zero(T))
-end
-
-# avoid linear indexing?
-@propagate_inbounds function getindex(D::SDiagonal{N,T}, k::Int) where {N,T}
-    i, j = CartesianIndices(size(D))[k].I
-    D[i,j]
-end
+#@inline (::Type{SDiagonal{N,T}})(a::AbstractVector) where {N,T} = Diagonal(SVector{N,T}(a))
+@inline (::Type{SDiagonal{N,T}})(a::Tuple) where {N,T} = Diagonal(SVector{N,T}(a))
+@inline (::Type{SDiagonal{N}})(a::Tuple) where {N} = Diagonal(SVector{N}(a))
 
-ishermitian(D::SDiagonal{N, T}) where {N,T<:Real} = true
-ishermitian(D::SDiagonal) = all(D.diag .== real(D.diag))
-issymmetric(D::SDiagonal) = true
-isposdef(D::SDiagonal) = all(D.diag .> 0)
+SDiagonal(a::SVector) = Diagonal(a)
+SDiagonal(a::StaticMatrix{N,N,T}) where {N,T} = Diagonal(diag(a))
 
-factorize(D::SDiagonal) = D
+size(::Type{<:SDiagonal{N}}) where {N} = (N,N)
+size(::Type{<:SDiagonal{N}}, d::Int) where {N} = d > 2 ? 1 : N
 
-==(Da::SDiagonal, Db::SDiagonal) = Da.diag == Db.diag
--(A::SDiagonal) = SDiagonal(-A.diag)
-+(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag + Db.diag)
--(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag - Db.diag)
--(A::SDiagonal, B::SMatrix) = typeof(B)(I)*A - B
-
-*(x::T, D::SDiagonal) where {T<:Number} = SDiagonal(x * D.diag)
-*(D::SDiagonal, x::T) where {T<:Number} = SDiagonal(D.diag * x)
-/(D::SDiagonal, x::T) where {T<:Number} = SDiagonal(D.diag / x)
-*(Da::SDiagonal, Db::SDiagonal) = SDiagonal(Da.diag .* Db.diag)
-*(D::SDiagonal, V::AbstractVector) = D.diag .* V
-*(D::SDiagonal, V::StaticVector) = D.diag .* V
-*(A::StaticMatrix, D::SDiagonal) = scalem(A,D.diag)
-*(D::SDiagonal, A::StaticMatrix) = scalem(D.diag,A)
+# define specific methods to avoid allocating mutable arrays
+*(A::StaticMatrix, D::SDiagonal) = A .* D.diag'
```

Comment:
Would be good to time this against `scalem`

Reply:
Hmm, it does seem to be about 10x slower for a 4x4 matrix. I wonder why it's so much slower?

Reply:
ah, I forgot to use `$` interpolation. They're more or less the same.